### ISSUE TITLE
feat(optimizers): Add AdamW — decoupled weight decay optimizer

### ISSUE THREAD

**Issue by sanskarpan** [sanskarpan/Optimization#26]
## Description
Implement **AdamW** (Loshchilov & Hutter, 2019) as a new optimizer class.

## Algorithm
Weight decay is decoupled from the gradient update:
```
θ ← θ * (1 - lr * weight_decay)   # decoupled decay
m = β₁*m + (1-β₁)*g
v = β₂*v + (1-β₂)*g²
θ ← θ - lr * m̂ / (√v̂ + ε)
```
Unlike L2-regularized Adam, weight decay is NOT adapted by the second moment.

## Acceptance Criteria
- [ ] `AdamW` class with `learning_rate`, `beta1`, `beta2`, `eps`, `weight_decay` params
- [ ] `weight_decay=0` produces identical results to Adam
- [ ] `get_state()`, `load_state()`, `__repr__()` implemented
- [ ] NaN/Inf guard in `step()`
- [ ] Tests: decoupled decay, zero-decay equals Adam, convergence, repr
